cmd/crowdsec/serve.go

```diff
@@ -1,78 +1,131 @@
 package main
 
 import (
-	"fmt"
 	"os"
-	"syscall"
 	"time"
 
+	"github.com/crowdsecurity/crowdsec/pkg/acquisition"
+	leaky "github.com/crowdsecurity/crowdsec/pkg/leakybucket"
 	"github.com/crowdsecurity/crowdsec/pkg/outputs"
 	log "github.com/sirupsen/logrus"
 
 	"github.com/sevlyar/go-daemon"
 )
 
+//debugHandler is kept as a dev convenience : it shuts down and serialize internal state
+func debugHandler(sig os.Signal) error {
+	var tmpFile string
+	var err error
+	//stop go routines
+	if err := ShutdownRoutines(); err != nil {
+		log.Warningf("Failed to shut down routines: %s", err)
+	}
+	//todo : properly stop acquis with the tail readers
+	if tmpFile, err = leaky.DumpBucketsStateAt(time.Now(), buckets); err != nil {
+		log.Warningf("Failed dumping bucket state : %s", err)
+	}
+	if err := leaky.ShutdownAllBuckets(buckets); err != nil {
+		log.Warningf("while shutting down routines : %s", err)
+	}
+	log.Printf("shutdown is finished buckets are in %s", tmpFile)
+	return nil
+}
+
 func reloadHandler(sig os.Signal) error {
-	dumpMetrics()
+	var tmpFile string
+	var err error
+	//stop go routines
+	if err := ShutdownRoutines(); err != nil {
+		log.Fatalf("Failed to shut down routines: %s", err)
+	}
+	if tmpFile, err = leaky.DumpBucketsStateAt(time.Now(), buckets); err != nil {
+		log.Fatalf("Failed dumping bucket state : %s", err)
+	}
+
+	if err := leaky.ShutdownAllBuckets(buckets); err != nil {
+		log.Fatalf("while shutting down routines : %s", err)
+	}
+	//reload all and start processing again :)
+	if err := LoadParsers(cConfig); err != nil {
+		log.Fatalf("Failed to load parsers: %s", err)
+	}
+
+	if err := LoadBuckets(cConfig); err != nil {
+		log.Fatalf("Failed to load scenarios: %s", err)
+
+	}
+	//restore bucket state
+	log.Warningf("Restoring buckets state from %s", tmpFile)
+	if err := leaky.LoadBucketsState(tmpFile, buckets, holders); err != nil {
+		log.Fatalf("unable to restore buckets : %s", err)
+	}
+
+	if err := LoadOutputs(cConfig); err != nil {
+		log.Fatalf("failed to initialize outputs : %s", err)
+	}
+
+	if err := LoadAcquisition(cConfig); err != nil {
+		log.Fatalf("Error while loading acquisition config : %s", err)
+	}
+	//Start the background routines that comunicate via chan
+	log.Infof("Starting processing routines")
+	inputLineChan, err := StartProcessingRoutines(cConfig)
+	if err != nil {
+		log.Fatalf("failed to start processing routines : %s", err)
+	}
+
+	//Fire!
+	log.Warningf("Starting processing data")
+
+	acquisition.AcquisStartReading(acquisitionCTX, inputLineChan, &acquisTomb)
+
+	log.Printf("Reload is finished")
+	//delete the tmp file, it's safe now :)
+	if err := os.Remove(tmpFile); err != nil {
+		log.Warningf("Failed to delete temp file (%s) : %s", tmpFile, err)
+	}
 	return nil
 }
 
-func termHandler(sig os.Signal) error {
-	log.Warningf("Shutting down routines")
+func ShutdownRoutines() error {
+	var reterr error
```

Comment:
maybe we want an array of `error` no ?

if multiple Tomb failed to Kill, the error that will be printed will concern only the last failing one

Reply:
I think single error is better, but I'm going to print the errors on the fly :)